src/provider/controller/IntentHandler.ts

```diff
@@ -109,21 +111,29 @@ export class IntentHandler {
     }
 
     private async fireIntent(intent: Intent, appInfo: Application): Promise<IntentResolution> {
-        const appWindows = await this._model.findOrCreate(appInfo);
-        // to decide between focus nothing or apps with intent listener
-        const dispatchResults = await Promise.all(appWindows.map(async (window: AppWindow): Promise<boolean> => {
-            if (await window.isReadyToReceiveIntent(intent.type)) {
-                const payload: ReceiveIntentPayload = {context: intent.context, intent: intent.type};
-
-                // TODO: Implement a timeout so a misbehaving intent handler can't block the intent raiser (SERVICE-555)
-                await this._apiHandler.dispatch(window.identity, APIToClientTopic.RECEIVE_INTENT, payload);
-                return true;
-            } else {
+        await this._model.ensureRunning(appInfo);
+
+        // TODO: Revisit timeout logic [SERVICE-556]
+        let dispatchResults: [boolean, boolean[] | undefined] | undefined = undefined;
+        dispatchResults = await withTimeout(Timeouts.ADD_INTENT_LISTENER, (async () => {
+            const appWindows = await this._model.expectWindowsForApp(appInfo);
+
+            // Wait for windows to add intent listener, then dispatch payload
+            return Promise.all(appWindows.map(async (window: AppWindow): Promise<boolean> => {
+                if (await window.isReadyToReceiveIntent(intent.type)) {
+                    const payload: ReceiveIntentPayload = {context: intent.context, intent: intent.type};
+
+                    // TODO: Implement a timeout so a misbehaving intent handler can't block the intent raiser [SERVICE-555]
+                    if (dispatchResults === undefined) {
```

Comment:
Just to get my head around this one - When would dispatchResults be `undefined` if we are setting it be the promise this is called from?